tests/unit/coordinator/plugins/clickhouse/test_parts.py

```diff
@@ -52,26 +56,55 @@ def create_part_from_part_files(part_files: List[SnapshotFile]) -> Part:
 TABLE_1_PART_3 = create_part_files(table_uuid=T1_UUID, part_name="all_2_2_0", digest_seed="same")
 TABLE_3_PART_1A = create_part_files(table_uuid=T3_UUID, part_name="all_0_0_0", digest_seed="one")
 TABLE_3_PART_1B = create_part_files(table_uuid=T3_UUID, part_name="all_0_0_0", digest_seed="other")
+DATABASES_REPLICAS = [
+    DatabaseReplica(shard_name="shard_1", replica_name="node_1"),
+    DatabaseReplica(shard_name="shard_1", replica_name="node_2"),
+]
+SHARDED_DATABASE_REPLICAS = [
+    DatabaseReplica(shard_name="shard_1", replica_name="node_1"),
+    DatabaseReplica(shard_name="shard_2", replica_name="node_2"),
+]
 
 
 def test_group_files_into_parts_collects_parts_from_selected_tables() -> None:
     first_server_files = copy.deepcopy([*TABLE_1_PART_1, *TABLE_1_PART_2, *TABLE_1_PART_3])
     second_server_files = copy.deepcopy([*TABLE_1_PART_1, *TABLE_1_PART_2, *TABLE_1_PART_3])
-    parts, other_files = group_files_into_parts([first_server_files, second_server_files], {T1_UUID, T2_UUID})
+    tables_replicas = {T1_UUID: DATABASES_REPLICAS, T2_UUID: DATABASES_REPLICAS}
+    parts, other_files = group_files_into_parts([first_server_files, second_server_files], tables_replicas)
     assert len(parts) == 3
     for part, original_files in zip(parts, [TABLE_1_PART_1, TABLE_1_PART_2, TABLE_1_PART_3]):
-        assert sorted(part.files.keys()) == sorted([original_file.relative_path for original_file in original_files])
-        for file_path, file in part.files.items():
-            assert file_path == file.snapshot_file.relative_path
-            assert file.servers == {0, 1}
+        assert part.snapshot_files == original_files
+        assert part.total_size == 1120
+        assert part.servers == {0, 1}
+    assert other_files[0] == []
+    assert other_files[1] == []
+
+
+def test_group_files_into_parts_does_not_merge_parts_across_shards() -> None:
+    first_server_files = copy.deepcopy([*TABLE_1_PART_1, *TABLE_1_PART_2, *TABLE_1_PART_3])
```

Comment:
is it worth it to write a test case when first_server_files != second_server_files? as far as I understand, it's sort-of covered in the `test_add_file_to_parts_fails_on_inconsistent_file` case, does it make sense to have a higher-level test for that as well?

Reply:
There are 2 different cases to tests :
 - a single part is present on multiple servers but some files within that part are missing on some server: this is very bad and means some of our assumption about how Clickhouse stores data: it's `test_add_file_to_parts_fails_on_inconsistent_file`
 - a entire part is present on one of the replica but not another (in the same shard): this is fine, the data snapshot happened before replication: this one is not tested but it should :+1: